crates/astria-core/src/primitive/v1/mod.rs

```diff
@@ -243,87 +247,238 @@ pub struct IncorrectRollupIdLength {
+    prefix: bech32::Hrp,
+}
+
+impl Address {
+    #[must_use = "the builder must be used to construct an address to be useful"]
+    pub fn builder() -> AddressBuilder {
+        AddressBuilder::new()
     }
 
     #[must_use]
-    pub const fn from_array(array: [u8; ADDRESS_LEN]) -> Self {
-        Self(array)
+    pub fn bytes(self) -> [u8; ADDRESS_LEN] {
+        self.bytes
     }
 
+    /// Convert a string containing a bech32m string to an astria address.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// + `input` is not bech32m encoded.
+    /// + the decoded data contained in `input` is not 20 bytes long.
+    /// + the bech32 hrp prefix exceeds 16 bytes.
+    pub fn try_from_bech32m(input: &str) -> Result<Self, AddressError> {
+        let (hrp, bytes) = bech32::decode(input).map_err(AddressError::bech32m_decode)?;
+        Self::builder()
+            .slice(bytes)
+            .prefix(hrp.as_str())
+            .try_build()
+    }
+
+    /// Convert [`Address`] to a [`raw::Address`].
+    // allow: panics are checked to not happen
+    #[allow(clippy::missing_panics_doc)]
     #[must_use]
     pub fn to_raw(&self) -> raw::Address {
+        let bech32m = bech32::encode_lower::<bech32::Bech32m>(self.prefix, &self.bytes())
+            .expect("must not fail because len(prefix) + len(bytes) <= 63 < BECH32M::CODELENGTH");
+        // allow: the field is deprecated, but we must still fill it in
+        #[allow(deprecated)]
         raw::Address {
-            inner: self.to_vec().into(),
+            inner: Bytes::new(),
+            bech32m,
         }
     }
 
     #[must_use]
     pub fn into_raw(self) -> raw::Address {
-        raw::Address {
-            inner: self.to_vec().into(),
-        }
+        self.to_raw()
     }
 
     /// Convert from protobuf to rust type an address.
     ///
     /// # Errors
     ///
     /// Returns an error if the account buffer was not 20 bytes long.
-    pub fn try_from_raw(raw: &raw::Address) -> Result<Self, IncorrectAddressLength> {
-        Self::try_from_slice(&raw.inner)
+    pub fn try_from_raw(raw: &raw::Address) -> Result<Self, AddressError> {
+        // allow: `Address::inner` field is deprecated, but we must still check it
+        #[allow(deprecated)]
+        let raw::Address {
+            inner,
+            bech32m,
+        } = raw;
+        if bech32m.is_empty() {
+            return Self::builder()
+                .slice(inner.as_ref())
+                .prefix(ASTRIA_ADDRESS_PREFIX)
+                .try_build();
+        }
+        if inner.is_empty() {
+            return Self::try_from_bech32m(bech32m);
+        }
+        Err(AddressError::fields_are_mutually_exclusive())
     }
 }
 
 impl AsRef<[u8]> for Address {
     fn as_ref(&self) -> &[u8] {
-        &self.0
+        &self.bytes
     }
 }
 
-impl From<[u8; ADDRESS_LEN]> for Address {
-    fn from(inner: [u8; ADDRESS_LEN]) -> Self {
-        Self(inner)
+// impl From<[u8; ADDRESS_LEN]> for Address {
```

Comment:
remove? i feel like it's kind of nice to have this though, using the astria prefix as default, but i get why you removed it

Reply:
Ah, commented code strikes again.

I agree, it's nice to have it, but it's so easy to get wrong - and then out of a sudden you have the wrong prefix in the code.

We could provide utility constructors like `Address::astria_from_array` or something like that - but not in this PR.